crypto/rand/rand_unix.c

```diff
@@ -263,6 +278,96 @@ int syscall_random(void *buf, size_t buflen)
     return -1;
 }
 
+#if !defined(OPENSSL_RAND_SEED_NONE) && defined(OPENSSL_RAND_SEED_DEVRANDOM)
+static const char *devrandom_paths[] = { DEVRANDOM };
+static int devrandom_files[OSSL_NELEM(devrandom_paths)];
+static int devrandom_close = 0;
+
+/*
+ * Verify that the FD associated with the random source is an open character
+ * device.
+ */
```

Comment:
Just a suggestion:

```C
/*
 * Verify that the file descriptor associated with the random source is
 * still valid and refers to a character device. The rationale for doing
 * this is the fact that it is not uncommon for daemons to close all open
 * file handles when daemonizing. So the handle might have been closed or
 * even reused for opening another file.
 */
```

Reply:
@EmericBr, would you mind checking whether nginx closes all open handles when chrooting? Because this would mean that this pull request would not work for your issue #5330 as expected (see [my previous post](https://github.com/openssl/openssl/pull/6432#discussion_r193962136)).

Reply:
See also https://github.com/openssl/openssl/pull/6432#discussion_r193815951

Reply:
Yes, that would be a problem :(

Additionally and mostly as a reminder to myself, I need to check that Linux on the S390 uses character devices for its alternative random seed device nodes.  I think other Unix based operating systems do, but I'll also check what I've got available too.

Reply:
Please don't use `devrandom` in the names.  Sure, this is *currently* very Unix centric, but should this become implemented on other platforms as well, `devrandom` is as foreign a name as it gets.

Reply:
The alternative I suggested was to cache the result of fstat() and then check that the st_dev, st_ino, st_mode and st_rdev fields are the same.

Reply:
Where does it say that nginx does chroot(2)? And if doesn't (I don't see that it does), how does any of this help? Note that I'm not saying that there is no application that doesn't do chroot(2), only that it's only appropriate if things are done for right reasons. Closing all file descriptors aside, question also is if we expect application to trust libcrypto with initialization before chroot(2)-ing? BTW, how many will recognize what does "(2)" after "chroot(2)" actually entail? ... Elsewhere it was effectively said "nothing will suit legacy application." Well, this suggestion won't suit arbitrary application, only one that does things in specific order. So question also is if implied legacy application[s] actually do things in the said order. Explicitly naming the application would help...

Reply:
> Where does it say that nginx does chroot(2)? 

`nginx` was mentioned in the context of #5330. I just checked it again, and it's not quite as I recalled:

It was not the original reporter @EmericBr who referred to it (sorry for the confusion), it was @lukastribus who mentioned it in https://github.com/openssl/openssl/issues/5330#issuecomment-391463231. So maybe Lukas can say more about it.

Reply:
> question also is if we expect application to trust libcrypto with initialization before chroot(2)-ing?

This brings up the interesting question whether libcrypto 1.1.0 would be able to initialize its 'old' eay random generator after a chroot and without having /dev/random available. If not, then at least this case would not be a regression.

Reply:
There are really only a number of different order you can do the various steps to become a daemon. https://stackoverflow.com/questions/17954432/creating-a-daemon-in-linux has an example that doesn't cover chrooting or changing uid.

Note somewhere in that sequence you can call an openssl function that initialized the RNG. This might be intentional, like calling it before chroot because you know it's not going to be available after chroot. And you still need to be root before you chroot.

Reply:
Ok, thanks for refreshing my memory:-)

Reply:
I'm not actually sure what our old code does fork(), do it just depend on getpid() being different, or does it try to read from /dev/urandom again after a fork?

Reply:
> Note somewhere in that sequence you can call an openssl function that initialized the RNG. This might be intentional, like calling it before chroot because you know it's not going to be available after chroot.

Though question is not what application *can* do, but what unnamed legacy application *does* (or doesn't, or "refuses" to) do.

> And you still need to be root before you chroot.

Exactly the point behind question if we [are in position to] expect application to trust libcrypto with initialization before chroot(2)-ing. I'd say it's only sensible to *not* expect that, in which case it would be more appropriate to give application the ultimate responsibility and provide interface for passing open file descriptor. This is opposite to keeping it open by library and providing interface for controlling it.

You can sense contradiction between two paragraphs. Indeed, on one hand we speak of legacy applications (where keeping /dev/[u]random open is probably the only possibility), and then it's suggested to not keep file descriptor (but count on application to pass one). The contradiction is resolved by additional logical clause, "**and** provide interface for controlling it". But there is additional implicit condition: **if** it turns out that keeping file descriptor open helps unnamed legacy application, **then** (and only then) it would be appropriate to keep /dev/[u]random open. But without providing interface to control it (as legacy application won't use anyway). In which case interface that allows to pass file descriptor would override the file descriptor libcrypto kept upon initialization. I mean one would have to close the initial one and switch to one provided by application.

To summarize. Does keeping /dev/random open **actually** help unnamed legacy application? If not, it's hardly a point to do that. As for non-legacy applications, I'd rather suggest interface for passing open file descriptor.

Reply:
There is at least one other named application, if you really care about putting names on it, and that's haproxy. But there are so many applications that we really don't know what they all do, or how it's common to set it up, that I think we should at least have a way to get them working. And I would prefer that we get most of them working by default, but I'm also sure that we won't be able to get them all working by default.

Reply:
The most likely applications where people will run into this problem is that they want to use TLS 1.3 on an older system and manually upgrade to a recent openssl. So things doing https are probably the most likely to break.

Reply:
> and provide interface for controlling it

Providing an API to configure the seed sources and specify the seed policy (enable or disable particular seed sources, set the seeding order, configure whether /dev/random should be kept open, etc.) would be a good idea. The idea has been discussed as far back as #4394, but we never got around to do anything about it. Currently, it is definitely post-1.1.1.

>  if it turns out that keeping file descriptor open helps unnamed legacy application

You are right, we should not add a workaround to an issue for which we don't even have concrete evidence that it is a real issue out there in userland. It is good to be prepared, so we should keep on discussing this pull request. But we still have the option to hold this pull request back until after the final release when we have more concrete feedback about how many applications are affected and whether it is really needed.

Reply:
I've changed to the st_dev, st_ino, st_mode and st_rdev check method.  It's altogether better thanks.

Reply:
I'm not a great fan of having the application pass an FD.  There is a lot of nonsense regarding the gathering and use of seed data.  Better that we, who have at least some idea of what to do, try to do the best that we can for the application.

As for the question of this being required or not or being post 1.1.1, I'm fairly ambivalent.  Having to have a random device in a `chroot(2)` area doesn't seem like a big problem.  That we're different from before could be more of a concern.

Reply:
Clarification: this is not about an unnamed application. This is about haproxy.

Emeric is the maintainer of the SSL subsystem in haproxy, I myself am working on haproxy as well. I reproduced this first-hand in haproxy.

In the other thread I linked to the following haproxy troubleshooting session (where Sander, Emeric and myself confirmed the issue):
[Haproxy 1.8 with OpenSSL 1.1.1-pre4 stops working after 1 hour](https://www.mail-archive.com/haproxy@formilux.org/msg29592.html)

Sander runs an older kernel without getrandom functionality, Emeric cross-compiles without kernels headers (so the raw getrandom syscall doesn't work), and I had the kernel but not the libc, so it started working for me only with the raw getrandom syscall feature in -pre6.


I can see now that the nginx issue was not caused by chroot(), but by a permission problem (after nginx downgraded privileges). I'd argue that there's a possibility that keeping the FD open would have helped the nginx case as-well, depending on the initialization sequence - but let's focus on what we actually know. I apologize for the confusion my nginx comment caused.


> Even more specifically, chroot(2) has to be called after rand is initialized.

Haproxy [deliberately calls RAND_bytes() before chroot()ing](https://github.com/haproxy/haproxy/commit/383085f6c0823069afd84411be28d5d5a3fe8e4e) to initialize OpenSSL's random number generator. Otherwise this would never have worked in the first place if I understand correctly.


> As for the question of this being required or not or being post 1.1.1

In my opinion, if this doesn't hit 1.1.1, the entire discussion is quite useless, considering the breakage, the fact that 1.1.1 will be LTS and the post 1.1.1 release schedule ... unless you are willing to backport such a change to 1.1.1 post release?

Reply:
> This is about haproxy.

Cool! Thanks for clarification!

Reply:
> unless you are willing to backport such a change to 1.1.1 post release?

My argumentation in https://github.com/openssl/openssl/pull/6432#discussion_r194250817 was to hold back the pull request as long as there is no concrete evidence of a regression yet, but of course keeping the option to backport it as bugfix to 1.1.1 if this would become necessary after the creation of the stable branch. Now since it has been clarified by you that haproxy is indeed affected by this regression, I would say it counts as bugfix and there is no need to hold back the pr.

Reply:
What about adding this but with the default being not keeping the FD open?
Since haproxy is already making an effort to start the RAND system, would it be possible to modify it to support the new API to maintain the FDs open?

Reply:
I guess making it not work by default in this instance isn't ideal.